contributing/development/core_and_modules/internal_rendering_architecture.rst

```diff
@@ -0,0 +1,762 @@
+disk pattern which allows for changing the number of samples and smoothly
+changing the quality.
+
+Godot also supports percentage-closer soft shadows (PCSS) for more realistic
+shadow penumbra rendering. PCSS shadows are limited to the Forward+
+backend as they're too demanding to be usable in the Forward Mobile backend.
+PCSS also uses a vogel-disk shaped kernel.
+
+Additionally, both shadow-mapping techniques rotate the kernel on a per-pixel
+basis to help soften under-sampling artifacts.
+
+The Compatibility backend doesn't support shadow mapping for any light types yet.
+
+Temporal antialiasing
+^^^^^^^^^^^^^^^^^^^^^
+
+.. note::
+
+  Only available in the Forward+ backend, not the Forward Mobile or
+  Compatibility methods.
+
+Godot uses a custom TAA implementation based on the old TAA implementation from
+`Spartan Engine <https://github.com/PanosK92/SpartanEngine>`__.
+
+Temporal antialiasing requires motion vectors to work. If motion vectors
+are not correctly generated, ghosting will occur when the camera or objects move.
+
+Motion vectors are generated on the GPU in the main material shader. This is
+done by running the vertex shader corresponding to the previous rendered frame
+(with the previous camera transform) in addition to the vertex shader for the
+current rendered frame, then storing the difference between them in a color buffer.
+
+Using `FSR 2.0 <https://github.com/GPUOpen-Effects/FidelityFX-FSR2>`__ instead
+of a custom TAA implementation is planned in a future release.
+
+**TAA resolve:**
+
+- `servers/rendering/renderer_rd/shaders/effects/taa_resolve.glsl <https://github.com/godotengine/godot/blob/master/servers/rendering/renderer_rd/shaders/effects/taa_resolve.glsl>`__
+
+Global illumination
+^^^^^^^^^^^^^^^^^^^
+
+.. note::
+
+  VoxelGI and SDFGI are only available in the Forward+ backend, not the
+  Forward Mobile or Compatibility methods.
+
+  LightmapGI *baking* is only available in the Forward+ and Forward
+  Mobile methods, and can only be performed within the editor (not in an
+  exported project). LightmapGI *rendering* is available in all methods
+  including Compatibility.
+
+Godot supports voxel-based GI (VoxelGI), signed distance field GI (SDFGI) and
+lightmap baking and rendering (LightmapGI). These techniques can be used
+simultaneously if desired.
+
+Lightmap baking happens on the GPU using Vulkan compute shaders – CUDA or OpenCL
+are **not** used. Using Vulkan compute allows for greater compatibility and
+performance on all kinds of GPUs. In contrast, CUDA only runs on NVIDIA hardware
+and OpenCL performance is crippled on NVIDIA GPUs.
+
+The GPU-based lightmapper is implemented in the LightmapperRD class, which inherits
+from the Lightmapper class. This allows for implementing additional lightmappers,
+paving the way for a future port of the CPU-based lightmapper present in Godot 3.x.
+This would allow baking lightmaps while using the Compatibility backend.
```

Comment:
It could be worth adding a brief paragraph on why Vulkan compute is used instead of CUDA/OpenCL (which might seem like the logical choices for doing this at first), but I think shortening this makes sense.

Reply:
Well, the main reason is just that we already have a vulkan context and it is way faster to use Vulkan compute then to take Vulkan GPU resources, upload them to CPU, download them back to GPU for CUDA, process on CUDA, upload back to CPU, put them in vulkan compatible format, download back to GPU for Vulkan.

That being said, in my opinion, a discussion on CUDA vs. Vulkan is not suitable for this document. CUDA isn't a viable choice for games. Our choice to not use it is not unique and does not require an explanation